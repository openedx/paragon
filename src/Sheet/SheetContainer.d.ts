/* eslint-disable react/require-default-props */
/* eslint-disable react/no-unused-prop-types */
import * as React from 'react';

export interface SheetContainerProps {
  children: React.ReactNode;
}

// eslint-disable-next-line react/prefer-stateless-function
declare class SheetContainer extends React.Component<SheetContainerProps> {}

export default SheetContainer;
