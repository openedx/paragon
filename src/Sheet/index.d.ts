import React from 'react';

export declare const POSITIONS: {
    left: string;
    right: string;
    top: string;
    bottom: string;
};

export declare const VARIANTS: {
    light: string;
    dark: string;
};

export interface SheetProps {
    blocking?: boolean;
    children?: React.ReactNode;
    position?: keyof typeof POSITIONS;
    show?: boolean;
    onClose?: () => void;
    variant?: keyof typeof VARIANTS;
}

declare class Sheet extends React.Component<SheetProps> {}

export default Sheet;
