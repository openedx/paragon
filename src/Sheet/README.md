---
title: 'Sheet'
type: 'component'
components:
- Sheet
categories:
- Overlays
status: 'New'
designStatus: 'In Progress'
devStatus: 'In Progress'
notes: |
  Window-level edge-anchored content container with elevation shadow, with option to block content.
---

## Basic Usage

```jsx live

() => {
  const [blocking, setBlocking] = useState(false);
  const [dark, setDark] = useState(false);
  const [position, setPosition] = useState('bottom');
  const [show, setShow] = useState(false);

  const positions = ['left', 'right', 'top', 'bottom'];

  return (
    <>
      <DropdownButton
        id="position-dropdown-btn"
        onSelect={setPosition}
        title="Sheet Position"
      >
        {positions.map(position => (
          <Dropdown.Item eventKey={position}>{position}</Dropdown.Item>
        ))}
      </DropdownButton><br />
      <Button onClick={() => setShow(true)}>
        Show the Sheet
      </Button>{' '}
      <Button onClick={() => setBlocking(!blocking)}>
        {blocking ? "Disable": "Enable"} blocking content
      </Button>{' '}
      <Button onClick={() => setDark(!dark)}>
        Set {dark ? "Light": "Dark"} mode
      </Button>

      <Sheet
        position={position}
        show={show}
        blocking={blocking}
        variant={dark ? 'dark' : 'light'}
        onClose={() => setShow(false)}
      >
        This is a Sheet component <br />
        <Button
          onClick={() => setShow(false)}
          variant={dark ? 'inverse-primary' : 'primary'}
        >
          Hide Me!
        </Button>
      </Sheet>
    </>
  )
}
```
