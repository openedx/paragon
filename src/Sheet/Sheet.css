/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-elevation-sheet-zindex-main: 1032;
  --pgn-elevation-sheet-zindex-backdrop: 1031;
}
