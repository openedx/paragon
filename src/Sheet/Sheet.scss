.pgn__sheet-skrim {
  width: 100%;
  height: 100%;
  background-color: rgba(196, 196, 196, .5);
  position: fixed;
  top: 0;
  left: 0;

  &.hidden {
    display: none;
  }

  z-index: $zindex-sheet-backdrop;
}

%component-left {
  left: 0;
  box-shadow: 8px 0 16px 0 rgba(0, 0, 0, .15);
}

%component-right {
  right: 0;
  box-shadow: -8px 0 16px 0 rgba(0, 0, 0, .15);
}

.pgn__sheet-component-left {
  @extend %component-left;
}

.pgn__sheet-component-right {
  @extend %component-right;
}

.pgn__sheet-component {
  position: fixed;
  padding: 1.25rem;
  background-color: white;
  z-index: $zindex-sheet;

  &.pgn__sheet__dark {
    background-color: $dark-500;
    color: $light-300;
  }

  &.bottom {
    bottom: 0;
    box-shadow: 0 -8px 16px 0 rgba(0, 0, 0, .15);
  }

  &.top {
    top: 0;
    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, .15);
  }

  &.left {
    @extend %component-left;

    [dir="rtl"] & {
      @extend %component-right;

      left: auto;
    }
  }

  &.right {
    @extend %component-right;

    [dir="rtl"] & {
      @extend %component-left;

      right: auto;
    }
  }

  &.bottom,
  &.top {
    width: 100%;
    left: 0;
    max-height: 50%;
  }

  &.left,
  &.right {
    max-width: 272px;
    height: 100%;
    top: 0;
  }
}
