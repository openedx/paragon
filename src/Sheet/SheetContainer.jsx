import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

class SheetContainer extends React.Component {
  constructor(props) {
    super(props);
    this.sheetRootName = 'sheet-root';
    if (typeof document === 'undefined') {
      this.rootElement = null;
    } else if (document.getElementById(this.sheetRootName)) {
      this.rootElement = document.getElementById(this.sheetRootName);
    } else {
      const rootElement = document.createElement('div');
      rootElement.setAttribute('id', this.sheetRootName);
      rootElement.setAttribute('class', 'sheet-container');
      rootElement.setAttribute('data-testid', 'sheet-container');
      this.rootElement = document.body.appendChild(rootElement);
    }
  }

  render() {
    if (this.rootElement) {
      return ReactDOM.createPortal(
        this.props.children,
        this.rootElement,
      );
    }
    return null;
  }
}

SheetContainer.propTypes = {
  children: PropTypes.node.isRequired,
};

export default SheetContainer;
