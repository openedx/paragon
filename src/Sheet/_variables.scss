$sheet-skrim-bg: var(--pgn-sheet-skrim-bg) !default;
$sheet-skrim-component-box-shadow: var(--pgn-sheet-skrim-component-box-shadow) !default;
