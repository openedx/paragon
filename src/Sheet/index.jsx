import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { FocusOn } from 'react-focus-on';

import SheetContainer from './SheetContainer';

export const POSITIONS = {
  left: 'left',
  right: 'right',
  top: 'top',
  bottom: 'bottom',
};

export const VARIANTS = {
  light: 'light',
  dark: 'dark',
};

class Sheet extends React.Component {
  constructor(props) {
    super(props);

    this.wrapperRef = React.createRef();
    this.renderSheet = this.renderSheet.bind(this);
  }

  renderSheet() {
    const { children, position, variant } = this.props;
    return (
      <div
        className={classNames(
          'pgn__sheet-component',
          `pgn__sheet__${variant}`,
          position,
        )}
        role="alert"
        aria-live="polite"
        aria-atomic="true"
      >
        <div className="pgn__sheet-content">
          { children }
        </div>
      </div>
    );
  }

  render() {
    const {
      blocking,
      show,
      onClose,
    } = this.props;
    if (!show) {
      return null;
    }
    return (
      <SheetContainer>
        <div
          className={classNames(
            'pgn__sheet-skrim',
            { hidden: !blocking },
          )}
          role="presentation"
        />
        <FocusOn
          onClickOutside={blocking ? () => {} : onClose}
          onEscapeKey={blocking ? () => {} : onClose}
          shards={[this.wrapperRef]}
        >
          {this.renderSheet()}
        </FocusOn>
      </SheetContainer>
    );
  }
}

Sheet.propTypes = {
  /** specifies whether the sheet provides a click-blocking scrim */
  blocking: PropTypes.bool,
  /** an element rendered inside the sheet */
  children: PropTypes.node,
  /** a string designating the sheet's position on the window */
  position: PropTypes.oneOf([
    POSITIONS.left,
    POSITIONS.right,
    POSITIONS.top,
    POSITIONS.bottom,
  ]),
  /** Boolean used to control whether the Sheet shows. */
  show: PropTypes.bool,
  /** Specifies function that controls `show` value. */
  onClose: PropTypes.func,
  /** a string designating which version of the sheet to show (light vs dark) */
  variant: PropTypes.oneOf([VARIANTS.light, VARIANTS.dark]),
};

Sheet.defaultProps = {
  blocking: false,
  children: undefined,
  position: POSITIONS.bottom,
  show: true,
  onClose: () => {},
  variant: VARIANTS.light,
};

export default Sheet;
