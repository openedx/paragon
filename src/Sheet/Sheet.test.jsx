import React from 'react';
import renderer from 'react-test-renderer';
import Sheet, { POSITIONS, VARIANTS } from './index';

/* eslint-disable react/prop-types */
jest.mock('./SheetContainer', () => function SheetContainerMock(props) {
  const { children, ...otherProps } = props;
  return (
    <sheet-container {...otherProps}>
      {children}
    </sheet-container>
  );
});

jest.mock('react-focus-on', () => ({
  FocusOn: (props) => {
    const { children, ...otherProps } = props;
    return (
      <focus-on {...otherProps}>{children}</focus-on>
    );
  },
}));

const testContent = (<div className="sheet-content">Hi</div>);

const renderJSON = (jsxContent) => renderer.create(jsxContent).toJSON();

describe('<Sheet />', () => {
  describe('snapshots', () => {
    test('default args snapshot: bottom, show, light', () => {
      const el = renderJSON(<Sheet>{testContent}</Sheet>);
      expect(el).toMatchSnapshot();
    });

    test('blocking, left snapshot', () => {
      expect(
        renderJSON(<Sheet blocking position={POSITIONS.left} />),
      ).toMatchSnapshot();
    });

    test('dark, right snapshot', () => {
      expect(
        renderJSON(<Sheet position={POSITIONS.right} variant={VARIANTS.dark} />),
      ).toMatchSnapshot();
    });
  });
  describe('correct rendering', () => {
    it('returns empty render if show is false', () => {
      expect(renderJSON(<Sheet show={false} />)).toEqual(null);
      expect(renderJSON(<Sheet />)).not.toEqual(null);
    });
  });
});
