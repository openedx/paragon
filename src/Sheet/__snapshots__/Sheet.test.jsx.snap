// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sheet /> snapshots blocking, left snapshot 1`] = `
<sheet-container>
  <div
    className="pgn__sheet-skrim"
    role="presentation"
  />
  <focus-on
    onClickOutside={[Function]}
    onEscapeKey={[Function]}
    shards={
      Array [
        Object {
          "current": null,
        },
      ]
    }
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      className="pgn__sheet-component pgn__sheet__light left"
      role="alert"
    >
      <div
        className="pgn__sheet-content"
      />
    </div>
  </focus-on>
</sheet-container>
`;

exports[`<Sheet /> snapshots dark, right snapshot 1`] = `
<sheet-container>
  <div
    className="pgn__sheet-skrim hidden"
    role="presentation"
  />
  <focus-on
    onClickOutside={[Function]}
    onEscapeKey={[Function]}
    shards={
      Array [
        Object {
          "current": null,
        },
      ]
    }
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      className="pgn__sheet-component pgn__sheet__dark right"
      role="alert"
    >
      <div
        className="pgn__sheet-content"
      />
    </div>
  </focus-on>
</sheet-container>
`;

exports[`<Sheet /> snapshots default args snapshot: bottom, show, light 1`] = `
<sheet-container>
  <div
    className="pgn__sheet-skrim hidden"
    role="presentation"
  />
  <focus-on
    onClickOutside={[Function]}
    onEscapeKey={[Function]}
    shards={
      Array [
        Object {
          "current": null,
        },
      ]
    }
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      className="pgn__sheet-component pgn__sheet__light bottom"
      role="alert"
    >
      <div
        className="pgn__sheet-content"
      >
        <div
          className="sheet-content"
        >
          Hi
        </div>
      </div>
    </div>
  </focus-on>
</sheet-container>
`;
